### PR TITLE
[MINOR][DOC] Performance Test Documentation

### PR DESCRIPTION
**Pull request by krishnakalyan3**
#### Performance Test Documentation with the following sections

- [x] Architecture
- [x] Adding New Algorithms
- [x] Current Default Settings
- [x] Examples
- [x] Troubleshooting

### CODE REVIEW

**Comment by krishnakalyan3** [comment]
ping @nakul02. 
Could you please review this PR.

File: docs/python-performance-test.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+Following changes to `predict.py`:
+
+- Create the function with `algoname + _ + predict`.
+- This function needs to have the following arguments `save_file_name`, `datagen_dir`, `train_dir`, `predict_dir`.
+- Constants and arguments for the training script should be defined in function.
+- Test the perf test with the algorithm with `mode` as `predict`.
+- Check output folders, json files, output log.
+- Check for possible errors if these folders/files do not exist. (Please see the troubleshooting section).
+
```
Comment (nakul02) [inline docs/python-performance-test.md]: Include a note about algorithms that don't have `prediction`.

File: docs/python-performance-test.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+### Current Default Settings
+Default setting for our performance test in-case these arguments are not provided by the user to override them.
+
```
Comment (nakul02) [inline docs/python-performance-test.md]: +Default setting for our performance test in-case these arguments are not provided by the user to override them.` - this is sort of confusing. Just leave it to read `default settings`. The rest is obvious.

File: docs/python-performance-test.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+- Matrix size to 10,000 rows and 100 columns.
+- Execution mode `singlenode`.
+- Operation modes `data-gen`, `train` and `predict` in sequence.
+- Generate `dense` and `sparse` matrix. if `sparse` not available, generate only `dense` matrix.
+
```
Comment (nakul02) [inline docs/python-performance-test.md]: There is this semantic where if the data is present for sparse and the user specifies that he wants to run the perf test for only dense, his request is ignored.
You should either document this here or change the implementation to not do this.

File: docs/python-performance-test.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+`
+Run performance test for the algorithms `m-svm` with `multinomial` family. Run only data generation and training operations.
+
+`./scripts/perftest/python/run_perftest.py --family regression2 --filename new_log
+`
+Run performance test for all algorithms under the family `regression2` and log with filename `new_log`.
+
+### Operational Notes
+All performance test depend mainly on two scripts for execution `systemml-standalone.py` and `systemml-spark-submit.py`. Incase we need to change standalone or spark parameters we need to manually change these parameters in their respective scripts.
+
```
Comment (nakul02) [inline docs/python-performance-test.md]: Remember to update this when you add in the ability to specify spark & standalone parameters as part of phase 2.

File: docs/python-performance-test.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+MultiLogReg|predict|0|10k_100|dense|4.780
+
+These logs can be found in `temp` folder (`$SYSTEMML_HOME/scripts/perftest/temp`) in-case not overridden by `--temp-dir`. This `temp` folders also contain the data generated during our performance test.
+
+Every time a script executes in `data-gen` mode successfully, we write a `_SUCCESS` file. If this file exists we ensures that re-run of the same script is not possible as data already exists.
+
+### Troubleshooting
+We can debug the performance test by making changes in the following locations based on 
+
+- [utils.py][line 174] In  uncommenting the debug print statement in the function `exec_dml_and_parse_time`. This allows us to inspect the subprocess string being executed.
```
Comment (nakul02) [inline docs/python-performance-test.md]: Specifying line numbers is very brittle. Either point to a variable or a function where this change needs to happen. Preferably not a line number.

File: docs/python-performance-test.md (hunk 4 above)
Comment (krishnakalyan3) [inline docs/python-performance-test.md]: @nakul02 agree. Is it okay to keep this as WIP util end of phase 3?.

File: docs/python-performance-test.md (hunk 3 above)
Comment (krishnakalyan3) [inline docs/python-performance-test.md]: @nakul02 good catch!. I will fix this in the code.

**Comment by nakul02** [comment]
This is good work! Documentation is always appreciated.

Ping @asurve - this document should help with running the perf tests using @krishnakalyan3's perf test script.

File: docs/python-performance-test.md (hunk 4 above)
Comment (nakul02) [inline docs/python-performance-test.md]: sure.

**Comment by krishnakalyan3** [comment]
@nakul02 thanks for the review.

**Comment by nakul02** [comment]
@krishnakalyan3 - I would not expect you to keep this PR open till the end of GSoC. It should be in a good enough shape to be commit-able right now. With each phase, you should update it and commit it as part of that phase. 
Let me know if you think this is commit-able right now.

**Comment by krishnakalyan3** [comment]
@nakul02, yes its commit-able.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1726/

**Comment by nakul02** [comment]
Let me know when you are done.

**Comment by krishnakalyan3** [comment]
@nakul02 could you please review/merge this PR. I will update document as and when required.

**Comment by akchinSTC** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://sparktc.ibmcloud.com/jenkins/job/SystemML-PullRequestBuilder/1748/

**Comment by nakul02** [comment]
Thanks @krishnakalyan3 
I shall merge.

**Pull request closed without merging**